keanu-project/src/main/java/io/improbable/keanu/tensor/INDArrayShim.java

```diff
@@ -24,76 +24,116 @@
  */
 public class INDArrayShim {
 
-    public static void muli(INDArray left, INDArray right, INDArray result) {
+    public static INDArray muli(INDArray left, INDArray right) {
         if (Arrays.equals(left.shape(), right.shape())) {
-            left.muli(right);
+            return left.muli(right);
         } else {
-            broadcastMultiply(left, right, result);
+            return broadcastMultiply(left, right);
         }
     }
 
-    private static void broadcastMultiply(INDArray a, INDArray b, INDArray result) {
-        int[] broadcastDimensions = Shape.getBroadcastDimensions(a.shape(), b.shape());
-        execBroadcast(a, b,
-            new BroadcastMulOp(a, b, result, broadcastDimensions)
+    private static INDArray broadcastMultiply(INDArray a, INDArray b) {
+        int[] broadcastDimensions = getBroadcastDimensions(a.shape(), b.shape());
+        if (a.shape().length < b.shape().length) {
+            return broadcastMultiply(b, a);
+        }
+        return execBroadcast(a, b,
+            new BroadcastMulOp(a, b, a.dup(), broadcastDimensions)
         );
     }
 
-    public static void divi(INDArray left, INDArray right, INDArray result) {
+    public static INDArray divi(INDArray left, INDArray right) {
         if (Arrays.equals(left.shape(), right.shape())) {
-            left.divi(right);
+            return left.divi(right);
         } else {
-            broadcastDivide(left, right, result);
+            return broadcastDivide(left, right);
         }
     }
 
-    private static void broadcastDivide(INDArray a, INDArray b, INDArray result) {
-        int[] broadcastDimensions = Shape.getBroadcastDimensions(a.shape(), b.shape());
-        execBroadcast(a, b,
-            new BroadcastDivOp(a, b, result, broadcastDimensions)
+    private static INDArray broadcastDivide(INDArray a, INDArray b) {
+        int[] broadcastDimensions = getBroadcastDimensions(a.shape(), b.shape());
+        if (a.shape().length < b.shape().length) {
+            return broadcastMultiply(b, a.rdiv(1.0));
```

Comment:
Think you mean divide here?

Reply:
Actually, sorry - I completely mis-read this - please ignore.

Reply:
although perhaps we should just call the Divide function anyway to make it easier to read?